File: ImperatorToCK3/Source/Imperator/Families/Families.cpp
#include "Families.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



void Imperator::Families::loadFamilies(const std::string& thePath) {
	registerKeys();
	parseFile(thePath);
	clearRegisteredKeywords();
}


void Imperator::Families::loadFamilies(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::Families::removeUnlinkedMembers() {
	for (const auto& family : families) {
		family.second->removeUnlinkedMembers();
	}
}


void Imperator::Families::registerKeys() {
	registerRegex(commonItems::integerRegex, [this](const std::string& theFamilyID, std::istream& theStream) {
		const auto familyStr = commonItems::stringOfItem(theStream).getString();
		if (familyStr.find('{') != std::string::npos) {
			std::stringstream tempStream(familyStr);
			const auto ID = commonItems::stringToInteger<unsigned long long>(theFamilyID);
			std::shared_ptr<Family> newFamily = familyFactory.getFamily(tempStream, ID);
			auto [iterator, inserted] = families.emplace(newFamily->getID(), newFamily);
			if (!inserted) {
				Log(LogLevel::Debug) << "Redefinition of family " << theFamilyID;
				iterator->second = newFamily;
			}
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}



Imperator::FamiliesBloc::FamiliesBloc(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::FamiliesBloc::registerKeys() {
	registerKeyword("families", [this](std::istream& theStream) {
		families.loadFamilies(theStream);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Families/Families.cs
using commonItems;
using System.Collections.Generic;

namespace ImperatorToCK3.Imperator.Families {
	public class Families : Dictionary<ulong, Family> {
		public void LoadFamilies(string path) {
			var parser = new Parser();
			RegisterKeys(parser);
			parser.ParseFile(path);
		}
		public void LoadFamilies(BufferedReader reader) {
			var parser = new Parser();
			RegisterKeys(parser);
			parser.ParseStream(reader);
		}

		private void RegisterKeys(Parser parser) {
			parser.RegisterRegex(CommonRegexes.Integer, (reader, familyIdStr) => {
				var familyStr = new StringOfItem(reader).String;
				if (!familyStr.Contains('{')) {
					return;
				}
				var tempReader = new BufferedReader(familyStr);
				var id = ulong.Parse(familyIdStr);
				var newFamily = Family.Parse(tempReader, id);
				var inserted = TryAdd(newFamily.Id, newFamily);
				if (!inserted) {
					Logger.Debug($"Redefinition of family {familyIdStr}.");
					this[newFamily.Id] = newFamily;
				}
			});
			parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		}
		public void RemoveUnlinkedMembers() {
			foreach (var family in Values) {
				family.RemoveUnlinkedMembers();
			}
		}

		public static Families ParseBloc(BufferedReader reader) {
			var blocParser = new Parser();
			var families = new Families();
			blocParser.RegisterKeyword("families", reader =>
				families.LoadFamilies(reader)
			);
			blocParser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);

			blocParser.ParseStream(reader);
			blocParser.ClearRegisteredRules();

			Logger.Debug("Ignored Family tokens: " + string.Join(", ", Family.IgnoredTokens));
			return families;
		}
	}
}
